projects/plugins/jetpack/modules/masterbar/admin-menu/class-base-admin-menu.php

```diff
@@ -161,12 +171,47 @@ public function update_submenus( $slug, $submenus_to_update ) {
 			return;
 		}
 
-		foreach ( $submenu[ $slug ] as $i => $submenu_item ) {
-			if ( array_key_exists( $submenu_item[2], $submenus_to_update ) ) {
-				$submenu_item[2] = $submenus_to_update[ $submenu_item[2] ];
-				// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
-				$submenu[ $slug ][ $i ] = $submenu_item;
+		// This is needed for cases when the submenus to update have the same new slug.
+		$submenus_to_update = array_filter(
+			$submenus_to_update,
+			function ( $item, $old_slug ) {
+				return $item !== $old_slug;
+			},
+			ARRAY_FILTER_USE_BOTH
+		);
+
+		/**
+		 * Iterate over all submenu items and add the hide the submenus with CSS classes.
+		 * This is done separately of the second foreach because the position of the submenu might change.
+		 */
+		foreach ( $submenu[ $slug ] as $index => $item ) {
```

Comment:
Can we make a copy of the submenu before iterating it here, so we can merge the two `foreach` loops? I think arrays are assigned by copy and not by reference in PHP, so I guess something like below can work?

```php
$submenu_items = $submenu[ $slug ];
foreach ( $submenu_items as $index => $item ) {
  if ( ! array_key_exists( $item[2], $submenus_to_update ) ) {
    continue;
  }
  $this->hide_submenu_element( $index, $slug, $item );
  add_submenu_page(
    $slug,
    isset( $item[3] ) ? $item[3] : '',
    isset( $item[0] ) ? $item[0] : '',
    isset( $item[1] ) ? $item[1] : 'read',
    $submenus_to_update[ $item[2] ],
    '',
    $index
  );
}
```

Reply:
Yup, they are passed by value. However, this will not work because we are changing the position of the elements that should be hidden in $submenu global variable after the loop is executed the first time. Lets say we have this array:
```
[ 0 => 'item_with_old_url', 1 => 'item_with_old_url']
```
When the foreach will loop the first time, the first element will be hidden (as expected) and a new element will be added to the array - therefore, our  global submenu should look like this:
```
[ 0 => 'new_item_with_the_url_update', 1 => 'item_with_old_url', 2 => 'item_with_old_url']
```
When the foreach will loop again, the $index will move to 1 which for the copied submenu item would be correctly, however for the submenu from $submenu global variable will not be correct.

Reply:
Gotcha, the problem is that we'll be passing an `$index` to `hide_submenu_element` that is based on the copied submenu, not the actual global `$submenu`, and thus will hide an incorrect submenu in subsequent iterations. Thanks for clarifying it.